src/tests/end2end/SamplerTests.cpp

```diff
@@ -0,0 +1,195 @@
+
+namespace {
+    struct AddressModeTestCase {
+        nxt::AddressMode mMode;
+        uint8_t mExpected2;
+        uint8_t mExpected3;
+    };
+    AddressModeTestCase addressModes[] = {
+        nxt::AddressMode::Repeat,           0, 255,
+        nxt::AddressMode::MirroredRepeat, 255,   0,
+        nxt::AddressMode::ClampToEdge,    255, 255,
+    };
+}
+
+class SamplerTest : public NXTTest {
+protected:
+
+    template<typename T>
+    nxt::Buffer MakeVertexBuffer(std::vector<T> data) {
+        return utils::CreateFrozenBufferFromData(device, data.data(), static_cast<uint32_t>(data.size() * sizeof(T)), nxt::BufferUsageBit::Vertex);
+    }
+
+    void SetUp() override {
+        NXTTest::SetUp();
+        mFB = utils::CreateBasicFramebuffer(device, kRTSize, kRTSize);
+        mFB.color.TransitionUsage(nxt::TextureUsageBit::OutputAttachment);
+
+        mBindGroupLayout = device.CreateBindGroupLayoutBuilder()
+            .SetBindingsType(nxt::ShaderStageBit::Fragment, nxt::BindingType::Sampler, 0, 1)
+            .SetBindingsType(nxt::ShaderStageBit::Fragment, nxt::BindingType::SampledTexture, 1, 1)
+            .GetResult();
+
+        auto pipelineLayout = device.CreatePipelineLayoutBuilder()
+            .SetBindGroupLayout(0, mBindGroupLayout)
+            .GetResult();
+
+        auto vsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Vertex, R"(
+            #version 450
+            layout(location = 0) in vec4 pos;
+            void main() {
+                gl_Position = pos;
+            }
+        )");
+        auto fsModule = utils::CreateShaderModule(device, nxt::ShaderStage::Fragment, R"(
+            #version 450
+            layout(set = 0, binding = 0) uniform sampler sampler0;
+            layout(set = 0, binding = 1) uniform texture2D texture0;
+            layout(location = 0) out vec4 fragColor;
+
+            void main() {
+                fragColor = texture(sampler2D(texture0, sampler0), gl_FragCoord.xy / 2.0);
+            }
+        )");
+
+        auto inputState = device.CreateInputStateBuilder()
+             .SetAttribute(0, 0, nxt::VertexFormat::FloatR32G32B32A32, 0)
+             .SetInput(0, 4 * sizeof(float), nxt::InputStepMode::Vertex)
+             .GetResult();
+
+        mPipeline = device.CreateRenderPipelineBuilder()
+            .SetSubpass(mFB.renderPass, 0)
+            .SetLayout(pipelineLayout)
+            .SetStage(nxt::ShaderStage::Vertex, vsModule, "main")
+            .SetStage(nxt::ShaderStage::Fragment, fsModule, "main")
+            .SetInputState(inputState)
+            .GetResult();
+
+        auto texture = device.CreateTextureBuilder()
+            .SetDimension(nxt::TextureDimension::e2D)
+            .SetExtent(2, 2, 1)
+            .SetFormat(nxt::TextureFormat::R8G8B8A8Unorm)
+            .SetMipLevels(1)
+            .SetAllowedUsage(nxt::TextureUsageBit::TransferDst | nxt::TextureUsageBit::Sampled)
+            .GetResult();
+
+        RGBA8 data[64 * 2];
```

Comment:
nit: Also comment explaining what the created texture looks like.

Reply:
Done.